service/storage_proxy.cc

```diff
@@ -865,7 +865,11 @@ class storage_proxy::remote {
             std::move(pr), oda, rate_limit_info_opt, fence);
     }
 
-    future<> handle_truncate(rpc::opt_time_point timeout, sstring ksname, sstring cfname) {
+    future<> handle_truncate(rpc::opt_time_point timeout, sstring ksname, sstring cfname, std::optional<frozen_topology_guard> frozen_guard) {
+        std::unique_ptr<topology_guard> guard;
```

Comment:
but it will get destroyed as soon as `truncate_table_on_all_shards` returns a non-ready future since we don't pass the guard to it.
Should this handler be converted to a coroutine and use `co_await replica::database::truncate_table_on_all_shards`?
And then I'm not sure why we can't just instantiate a `topology_guard` rather than a `std::unique_ptr<topology_guard>`

Reply:
Yes, my mistake. The guard will be destroyed as soon as the function returns. I missed the fact that this is not a co-routine.

I used `std::unique_ptr<topology_guard>` to be able to handle cases where `frozen_guard` equals `std::nullopt`. `topology_guard` can only be instantiated with a `frozen_topology_guard` in the constructor. I think I could have instantiated it like this:
```c++
topology_guard guard(frozen_guard.value_or(frozen_topology_guard{}));
```
But I though the solution with the unique_ptr is a little more readable.

Reply:
std::unique_ptr mallocs its buffer while std::optional can use the stack/coroutine frame
so it would save an allocation

Reply:
Ok, I will change it to std::optional in the next version.

Reply:
Changed in v6